### DISCUSSION TITLE
Localization

### DISCUSSION THREAD

**Discussion by ghost** [mint-lang/mint#469] [General]
Hello! I've not tried using Mint yet, but I find it interesting. The problem is, as I understand, the Mint compiler generates JavaScript code for front-end application, one that runs in Node.js (pre-render) and another that runs in the web browser, or otherwise I'm wrong. I've used a frameworks (Sapper) that facilitate creating SPAs, but have had issues with asynchronous use of localization resources (imagine, multiple users rendering different languages at the same time).

I've a localization library ([link](https://github.com/recoyx/Localization-js)), which does not support collation, currency or date-time formatting (I'm waiting for ECMAScript Temporal proposal to be finished), but it does the general work. I think it'd be nice to integrate that in Mint. If that requires the programmer to integrate it, then the same issue would occur as when I was trying to use Sapper (SPA framework for Svelte).

**Comment by gdotdesign** [comment]
Hi :wave: 

> The problem is, as I understand, the Mint compiler generates JavaScript code for front-end application, one that runs in Node.js (pre-render) and another that runs in the web browser, or otherwise I'm wrong.

Mint generates code for running in the browser only it does not run in Node.js, also mint doesn't support pre-rendering yet.


> I've a localization library ([link](https://github.com/recoyx/Localization-js)), which does not support collation, currency or date-time formatting (I'm waiting for ECMAScript Temporal proposal to be finished), but it does the general work. I think it'd be nice to integrate that in Mint. If that requires the programmer to integrate it, then the same issue would occur as when I was trying to use Sapper (SPA framework for Svelte).

I'm not sure what you are referring to, you can interface with JavaScript using the methods described here https://www.mint-lang.com/guide/reference/javascript-interop. In your case you probably need to compile your library into a single JS file which you will be able to use in a Mint application.

**Comment by frnco** [comment]
Internationalization is a beast in itself, and as I see it Mint is still missing some key features to appropriately deal with that (Macros should solve this, I guess).

In my case, I've opted to use good ol' YAML to write localization files, and wrote a Ruby script to load the files and convert the values to Mint code, defining an `enum` with the available language codes, a `record Translation` that has all the available fields with their values being a String containing the actual text, and a `store I18n` which has a `state language` that tracks the selected language and a `get translation` that uses a `case` to return the translation for the language as defined by the `record`, meaning you just need to reference any field to get its translation.

This has the disadvantage of not knowing whether the code generated by ruby will work until I run it on mint, but on the other side, as soon as I run mint it will complain if any of the languages is missing even one field, since they won't match the `record` for the translation, and I didn't have any problems with that yet. I'm considering porting the Ruby code to Crystal to see if I can improve the type-checking side of things, but for now, in case anyone might find this useful, I thought I should share this.
